### PR TITLE
Hostile mobs

### PR DESCRIPTION
**Pull request by someaddons**
# Changes proposed in this pull request:
## Features:
- adding config option to make mobs hostile to citizens
- adding config option to let citizens and guards call for help when attacked
- citizens run away after getting hit and keeps avoiding the attacker(if there is one)
- reworking old avoidance AI, now only scans the area actively after beeing attacked before.
- adding mercenaries which defend the colony for a price
    - summoned at the townhall via a button
    - last 20 min and reappear to be hired after 40min(so a 20min cooldown)
- adding housing to guards in guardtowers

## Fixes:
- fix child sounds not beeing registered
- fix damage sources of our entites to be entitydamagesources (so attacked citizens recognise raiders correctly)
- fix raider attackspeed, previous formula was broken and not scaling at all
- fix raiders stuck recognition causing them to ignore entity collision
- fixing barracks citizen limit recalculation, was only recalculated on building upgrade instead of citizen assignment
- fixed two pathing bugs which caused entites to not move towards their target frequently, e.g. some raiders not moving to the colony.
- fix lumberjack trying to save null values to nbt

Review please

### CODE REVIEW

**Comment by OCDCleric** [comment]
I feel like if this gets implemented that it should be default: false or have 1 or 2 guards spawn in free for early game safety

**Comment by someaddons** [comment]
I think it is natural in our progression that at the start your citizens are vulnerable and require caretaking

**Comment by OCDCleric** [comment]
True, but this also means the prepatory period is probably going to be much longer. Worst case is that we start putting up walls around town hall first thing.

**Comment by someaddons** [comment]
Raiders happen realtively early aswell and are not preventable by lighting up the area, so should probably have both on the same setting

**Comment by OCDCleric** [comment]
Will the mobs trigger mourning if they kill citizens?

**Comment by Raycoms** [comment]
I would make it default false for the beginning and then turn it on after
we got some feedback on it.

Am So., 16. Juni 2019 um 12:56 Uhr schrieb someaddons <
notifications@github.com>:

**Comment by someaddons** [comment]
ye thats fine, we can also increase early defenses by e.g. giving the guard tower an additional guard, or letting it house the citizen

**Comment by OCDCleric** [comment]
Actually couldn't we have citizens fight? Such as;

1. If attacked search for Guards
 -If Yes, run away and call help
 -If No, Punch or Use Tools on hand to fight back.
2. If attacked search for Guards
 -If Yes, run away call for help
 -If No, call for help from nearby citizens and gang up on enemy mob.

**Comment by marchermans** [comment]
I think at least minimal self defence of citizens needs to be possible.

**Comment by someaddons** [comment]
ye I could add a small AI for the to fight back, want them to fight always, or just when no guards are close?

**Comment by Raycoms** [comment]
I'd say only when no guards are close, maybe even say that citizens within
a range come to help. Reminds me of aoe

**Comment by ArkenorOakshadow** [comment]
Could vary by class. The lumberjack has an axe, after all, so is better situated to put up a fight than, say, the baker.

**Comment by OCDCleric** [comment]
At worst could add a mercenary contract to town hall that creates patrolling mercs for a period of time?

**Comment by marchermans** [comment]
@OCDCleric That is a nice Idea.
@someaddons Could you take a look at that as well, spawn in temporary guards that can be bought in the TownHall?

**Comment by someaddons** [comment]
ye I could, would be kinda similar to raiders then. What kinda cost do you think one could use?

**Comment by someaddons** [comment]
So to sum it up:
- have mobs attack citizen
- citizen call guards and flee
- citizen fight back when no guards, calling nearby citizen
- guardtowers provide housing to guards as barracks
- Townhall option to spawn temporary protector guards

**Comment by marchermans** [comment]
Correct 😄  👍 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/3673/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

**Comment by marchermans** [comment]
Balance it a bit. I think a base cost, plus some cost per minecraft day?

**Comment by someaddons** [comment]
not sure which resource one could use as costs, my ideas would be:
- reduction in happiness when hiring the mercenaries
- mercenaries taking 70% of all new resources produced
- a direct material cost consumed when hiring, e.g. 100 wood
- a gold currency which we dont have yet :D

**Comment by marchermans** [comment]
Yeah, I was thinking something in that direction as well.

**Comment by OCDCleric** [comment]
1. Reductuon in happiness makes sense
2. What woul 70% of all new resources entail?
3. Material cost would be Food and Gold?
4. I sense a new worker for making currency.

Ps. Should probably make a different thread elsewhere to didcuss this lol.

File: src/api/java/com/minecolonies/api/configuration/Configurations.java [hunk 1]
```diff
@@ -113,6 +113,12 @@
         @Config.Comment("The minimum number of nights between raids")
         public int minimumNumberOfNightsBetweenRaids = 1;
 
+        @Config.Comment("Should Mobs attack citizens? default:false")
+        public boolean mobAttackCitizens = true;
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/configuration/Configurations.java]: default false, but here it is true? =D

File: src/api/java/com/minecolonies/api/entity/ai/pathfinding/AbstractWalkToProxy.java [hunk 2]
```diff
@@ -102,7 +102,7 @@ public boolean walkToBlock(@NotNull final BlockPos target, final int range, fina
                 currentProxy = target;
             }
 
-            this.resetProxyList();
+            proxyList.clear();
             return takeTheDirectPath(target, range, onMove);
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/entity/ai/pathfinding/AbstractWalkToProxy.java:106]: And I thought both things do the same thing =P

File: src/api/java/com/minecolonies/api/util/FireworkUtils.java [hunk 3]
```diff
@@ -22,15 +22,16 @@
     /**
      * Private constructor to hide the public one
      */
-    private FireworkUtils() {
+    private FireworkUtils()
+    {
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/util/FireworkUtils.java:26]: ty =D

File: src/main/java/com/minecolonies/coremod/client/render/mobs/RenderMercenary.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ * Renderer for EntityMercenary.
+ */
+public class RenderMercenary extends RenderBiped<EntityMercenary>
+{
+    /**
+     * Texture of the entity.
+     */
+    private static final ResourceLocation TEXTURE = new ResourceLocation("minecolonies:textures/entity/settlermale1.png");
+
+
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/client/render/mobs/RenderMercenary.java]: seems to be missing a bit of doc here

File: src/main/java/com/minecolonies/coremod/colony/Colony.java [hunk 5]
```diff
@@ -1563,4 +1572,31 @@ public void increaseBoughtCitizenCost()
...
+        return mercenaryLastUse;
+    }
+
+    /**
+     * Registers an entity related to this colony.
+     */
+    public void registerEntityWithColony(final IColonyRelatedEntity entity)
+    {
+        // Add entity registry logic here
+    }
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/Colony.java]: is there something missing here?

File: src/main/java/com/minecolonies/coremod/colony/ColonyView.java [hunk 6]
```diff
@@ -992,4 +1000,14 @@ public boolean isRaiding()
     {
         return this.horde > 0;
     }
+
+    public long getMercenaryLastUseTime()
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/ColonyView.java]: @Override

File: src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java [hunk 7]
```diff
@@ -364,6 +385,81 @@ public boolean attackEntityFrom(@NotNull final DamageSource damageSource, final
...
+        if (!(attacker instanceof EntityLivingBase))
+        {
+            moveAwayPath = this.getNavigator().moveAwayFromEntityLiving(this, 5, INITIAL_RUN_SPEED_AVOID);
+            return;
+        }
+
+        // Makes the avoidance AI take over.
+        currentlyFleeing = true;
+
+        if ((getCitizenJobHandler().getColonyJob() instanceof AbstractJobGuard))
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:407]: Should the guards also move away?

File: src/main/java/com/minecolonies/coremod/entity/ai/mobs/EntityMercenary.java [hunk 8]
```diff
@@ -0,0 +1,518 @@
...
+    private              Colony                   colony;
+    /**
+     * This entities minecolonies-Navigator.
+     */
+    private              PathNavigate             newNavigator;
+    /**
+     * Proxy for cheaper pathing.
+     */
+    private              GeneralEntityWalkToProxy proxy;
+    /**
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/entity/ai/mobs/EntityMercenary.java:68]: this formatting seems off.

File: src/api/java/com/minecolonies/api/configuration/Configurations.java (hunk 1 above)
Comment (someaddons) [reply src/api/java/com/minecolonies/api/configuration/Configurations.java]: oh forgot to change it back from the testing version :D

File: src/api/java/com/minecolonies/api/entity/ai/pathfinding/AbstractWalkToProxy.java (hunk 2 above)
Comment (someaddons) [reply src/api/java/com/minecolonies/api/entity/ai/pathfinding/AbstractWalkToProxy.java:106]: yea there is a subtle difference, which is that resetProxyList() also clears the currentProxy which just was set on the line before this

File: src/main/java/com/minecolonies/coremod/colony/Colony.java (hunk 5 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/colony/Colony.java]: oh thats just unused atm, not sure if we got a use for it yet. Probably should just remove it so we dont have new unused code

File: src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java (hunk 7 above)
Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:407]: guards wont move away, as they do not have the moveawayAI added at all.

File: src/main/java/com/minecolonies/coremod/colony/Colony.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/Colony.java]: Agreed

File: src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java (hunk 7 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:407]: moveAwayPath = this.getNavigator().moveAwayFromEntityLiving(this, 5, INITIAL_RUN_SPEED_AVOID)

but it does this for the guards, isn't that unnecessary overhead?

Comment (someaddons) [reply src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:407]: no this is intended, it is running away a little bit from the current pos when getting hurt by environmental, nonentity caused damage

**Comment by Raycoms** [comment]
Ahh okay

Am Sa., 13. Juli 2019 um 14:31 Uhr schrieb someaddons <
notifications@github.com>:

**Pull request merged**